File: 프로그래머스/1/12943. 콜라츠 추측/콜라츠 추측.js
// Use Recursion (Or use for loop for 500 times)
function collatz(num, cnt) {
    // Base case, Termination case
    if (num === 1) {
        return cnt;
    }
    if (cnt === 500) {
        return -1;
    }
    // Process
    if (num % 2 == 0) { // even
        return collatz(num/2, cnt + 1);
    } else {            // odd
        return collatz(num * 3 + 1, cnt + 1)
    }
}

function solution(num) {
    return collatz(num, 0);
}

File: 프로그래머스/1/12943. 콜라츠 추측/콜라츠 추측.py
# Use Recursion
def collatz(num, cnt):
    if num == 1:
        return cnt
    if cnt == 500:
        return -1
    if num % 2 == 0:    # even
        return collatz(num/2, cnt + 1)
    else:               # odd
        return collatz(num * 3 + 1, cnt + 1)

def solution(num):
    return collatz(num, 0)